File: clients/gm_download.f90
program gm_download
  use iso_fortran_env, only: error_unit
  character(4096) :: u, o, cmd, python
  character(64) :: os_name
  integer :: stat, env_stat
  logical :: windows
  if (command_argument_count() /= 2) then; write(error_unit,*) 'usage: gm_download URL OUTPUT'; stop 2; end if
  call get_command_argument(1,u); call get_command_argument(2,o)
  call get_environment_variable('OS', os_name, status=env_stat)
  windows = env_stat == 0 .and. trim(os_name) == 'Windows_NT'
  call get_environment_variable('GM_PYTHON', python, status=env_stat)
  if (env_stat /= 0 .or. len_trim(python) == 0) then
    if (windows) then; python = 'python'; else; python = 'python3'; end if
  end if
  if (windows) then
    if (index(trim(python)//trim(u)//trim(o), '"') /= 0) then
      write(error_unit,*) 'double quotes are not supported in arguments'; stop 2
    end if
    cmd = '""'//trim(python)//'" "clients/download.py" --url "'//trim(u)//'" --output "'//trim(o)//'""'
  else
    if (index(trim(python)//trim(u)//trim(o), "'") /= 0) then
      write(error_unit,*) 'single quotes are not supported in arguments'; stop 2
    end if
    cmd = "'"//trim(python)//"' 'clients/download.py' --url '"//trim(u)//"' --output '"//trim(o)//"'"
  end if
  call execute_command_line(trim(cmd), exitstat=stat)
  if (stat /= 0) then; write(error_unit,*) 'Python downloader failed'; stop 1; end if
end program gm_download

File: clients/gm_download.c
/* C entry point: gm_download <url> <output>. Integrity-aware tag downloads
   can use clients/download.py; this tiny ABI keeps C integrations portable. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef _WIN32
#include <process.h>
#else
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
#endif

int main(int n, char **v) {
  const char *python = getenv("GM_PYTHON");
  if (n != 3) {
    fprintf(stderr, "usage: gm_download URL OUTPUT\n");
    return 2;
  }
#ifdef _WIN32
  if (python == NULL || *python == '\0') python = "python";
  {
    char *url;
    char *output;
    intptr_t status;
    if (strchr(v[1], '"') != NULL || strchr(v[2], '"') != NULL) {
      fprintf(stderr, "double quotes are not supported in arguments\n");
      return 2;
    }
    url = malloc(strlen(v[1]) + 3);
    output = malloc(strlen(v[2]) + 3);
    if (url == NULL || output == NULL) {
      free(url);
      free(output);
      return 3;
    }
    sprintf(url, "\"%s\"", v[1]);
    sprintf(output, "\"%s\"", v[2]);
    status = _spawnlp(_P_WAIT, python, python, "clients/download.py",
                      "--url", url, "--output", output, NULL);
    free(url);
    free(output);
    if (status == -1) {
      perror("unable to start Python 3");
      return 127;
    }
    return status == 0 ? 0 : 1;
  }
#else
  if (python == NULL || *python == '\0') python = "python3";
  {
    pid_t child = fork();
    int status;
    if (child == -1) {
      perror("unable to fork");
      return 127;
    }
    if (child == 0) {
      execlp(python, python, "clients/download.py", "--url", v[1],
             "--output", v[2], (char *)NULL);
      perror("unable to start Python 3");
      _exit(127);
    }
    if (waitpid(child, &status, 0) == -1) {
      perror("unable to wait for Python 3");
      return 127;
    }
    return WIFEXITED(status) ? WEXITSTATUS(status) : 1;
  }
#endif
}
